kotlin/iwa-sample-app/src/main/java/com/esri/android/samples/iwasampleapp/MainActivity.kt

```diff
@@ -0,0 +1,173 @@
+        // Set authentication challenge handler
+        AuthenticationManager.setAuthenticationChallengeHandler(IWACustomChallengeHandler(this))
+        
+        // Sign in a portal
+        signinButton.setOnClickListener{
+            // Validate portal url
+            val url : String ? = portalUrl.text.toString()
+            if (url.isNullOrEmpty()) {
+                Toast.makeText(applicationContext, "Portal url is empty. Please enter portal url!", Toast.LENGTH_SHORT).show()
+                return@setOnClickListener
+            }
+            
+            // Clear credential cache. This is for test purpose. In a real app credential cache provides
+            // better user experience.
+            AuthenticationManager.CredentialCache.clear()
+            // Create a Portal object with loginRequired as true, which forces user to sign in
+            val portal : Portal = Portal(url, true)
+            portal.loadAsync()
+            portal.addDoneLoadingListener{
+                if (portal.loadStatus == LoadStatus.LOADED) {
+                    // Portal is loaded. Add more login here
+                    Toast.makeText(applicationContext, "Portal is loaded!", Toast.LENGTH_SHORT).show()
+                } else {
+                    // Portal is failed to load. Handler load errors
+                    val error = portal.loadError
+                    if (error.errorCode == 17) {
+                        // User canceled exception
+                        Toast.makeText(applicationContext, "Portal loading is cancelled by user!", Toast.LENGTH_SHORT).show()
+                    } else {
+                        // Other failures
+                        Toast.makeText(applicationContext, 
+                                "Portal fails to load: " + error.cause?.message ?: error.message, Toast.LENGTH_SHORT).show()
+                    }
+                }
+            }
+        }
+    }
+}
+
+const val MAX_ATTEMPTS = 5
+
+/**
+ * A custom authentication challenge handler to handle user credential challenge.
+ * 
+ * When a user credential challenge is issues a dialog will be presented for credential.
+ * The portal url will be displayed as message in the dialog. If wrong credential has been
+ * passed in the previous attempt, a different message will be display in the dialog.
+ * The dialog has two edit text boxes for username and password respectively. Other SDKs'
+ * sample may have one more parameter for IWA domain. As indicated by the Javadoc of UseCredential
+ * our SDK is in favor of passing username as username@domain. This sample doesn't provide
+ * a edit text box for domain to simplify the logic.
+ *  
+ * @see <a href="https://developers.arcgis.com/android/latest/api-reference/reference/com/esri/arcgisruntime/security/UserCredential.html#UserCredential(java.lang.String,%20java.lang.String)">Javadoc of UserCredential</a>
+ * 
+ */
+class IWACustomChallengeHandler(val activity: Activity) : AuthenticationChallengeHandler {
```

Comment:
put an empty line here for better readability

Reply:
added